kimchi/src/circuits/constraints.rs

```diff
@@ -343,35 +355,74 @@ impl<F: FftField + SquareRootField> LookupConstraintSystem<F> {
                     .chain(lookup_tables.into_iter())
                     .collect();
 
-                if lookup_tables.len() > 1 {
-                    panic!("Multiple lookup tables are currently not supported");
+                // Get the max width of all lookup tables
+                let max_table_width = lookup_tables
+                    .iter()
+                    .fold(0, |max_width, LookupTable { data, .. }| {
+                        std::cmp::max(max_width, data.len())
+                    });
+
+                let max_num_entries = d1_size - (ZK_ROWS as usize) - 1;
+
+                let mut lookup_table = vec![Vec::with_capacity(d1_size); max_table_width];
+                let mut table_ids: Vec<F> = Vec::with_capacity(d1_size);
+                let mut non_zero_table_id = false;
+                for table in lookup_tables.iter() {
+                    let table_len = table.data[0].len();
+
+                    // Update table IDs
+                    if table.id != 0 {
+                        non_zero_table_id = true;
+                    }
+                    let table_id: F = gate::i32_to_field(table.id);
+                    table_ids.extend((0..table_len).map(|_| table_id));
+
+                    // Update lookup_table values
+                    for (i, col) in table.data.iter().enumerate() {
+                        if col.len() != table_len {
+                            return Err(LookupError::InconsistentTableLength);
+                        }
+                        lookup_table[i].extend(col);
+                    }
+
+                    // Fill in any unused columns with 0
+                    for lookup_table in lookup_table.iter_mut().skip(table.data.len()) {
+                        lookup_table.extend((0..table_len).map(|_| F::zero()))
+                    }
+                }
+
+                // Note: we use `>=` here to leave space for the dummy value.
+                if lookup_table[0].len() >= max_num_entries {
+                    return Err(LookupError::LookupTableTooLong {
+                        length: lookup_table[0].len(),
+                        maximum_allowed: max_num_entries - 1,
+                    });
```

Comment:
I'm confused with this one as well. max_num_entries is the size of d1 minus the rows where the grand product argument is not applied. This would be the zk_rows as well as the row before that (needed to check that the final value of the grand product is 1). It looks like max_num_entries is calculated this way, so why not `>` instead of `>=`?

Reply:
The dummy is now hard-coded as 0 to make the constraints cheaper. We use that final position to insert the zero dummy value, in case it isn't already present.